### ISSUE TITLE
EAS local iOS build with production profile fails

### ISSUE THREAD

**Issue by krystianhub** [expo/eas-cli#2449]
### Summary

My local iOS "production" builds started to fail on recent versions of Expo and EAS CLI.

I suspect it has something to do with my main entry file called "index.js" (package.json) - however, that wasn't a problem about 2 weeks ago on "production" profiles and isn't a problem today for latest versions of Expo/EAS if you target development profile with developmentClient: true.

### Managed or bare?

Managed

### Environment

expo-env-info 1.2.0 environment info:
    System:
      OS: macOS 14.5
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 22.4.0 - /usr/local/bin/node
      npm: 10.8.1 - /usr/local/bin/npm
    Managers:
      CocoaPods: 1.15.2 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 23.5, iOS 17.5, macOS 14.5, tvOS 17.5, visionOS 1.2, watchOS 10.5
    IDEs:
      Xcode: 15.4/15F31d - /usr/bin/xcodebuild
    npmGlobalPackages:
      eas-cli: 10.1.1
    Expo Workflow: managed
    
---

✔ Check Expo config for common issues
✔ Check package.json for common issues
✔ Check native tooling versions
✔ Check dependencies for packages that should not be installed directly
✔ Check for common project setup issues
✔ Check npm/ yarn versions
✔ Check for issues with metro config
✔ Check Expo config (app.json/ app.config.js) schema
✔ Check that native modules do not use incompatible support packages
✔ Check for legacy global CLI installed locally
✔ Check that native modules use compatible support package versions for installed Expo SDK
✔ Check that packages match versions required by installed Expo SDK

Didn't find any issues with the project!
    

### Error output

EAS local build with "production" profile fails with the following error message:

```
[RUN_FASTLANE] › Executing expobug » Bundle React Native code and images
[RUN_FASTLANE]     Bundler cache is empty, rebuilding (this may take a minute)
[RUN_FASTLANE] ▸ ** BUILD FAILED **
[RUN_FASTLANE] ▸ The following build commands failed:
[RUN_FASTLANE] ▸ 	PhaseScriptExecution Bundle\ React\ Native\ code\ and\ images /var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/eb840d6d-644a-4121-9f78-90d18b7a2f4e/build/ios/build/Build/Intermediates.noindex/expobug.build/Release-iphonesimulator/expobug.build/Script-00DD1BFF1BD5951E006B06BC.sh (in target 'expobug' from project 'expobug')
[RUN_FASTLANE] ▸ (1 failure)
[RUN_FASTLANE] ** BUILD FAILED **
[RUN_FASTLANE] 
[RUN_FASTLANE] 
[RUN_FASTLANE] The following build commands failed:
[RUN_FASTLANE] 	PhaseScriptExecution Bundle\ React\ Native\ code\ and\ images /var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/eb840d6d-644a-4121-9f78-90d18b7a2f4e/build/ios/build/Build/Intermediates.noindex/expobug.build/Release-iphonesimulator/expobug.build/Script-00DD1BFF1BD5951E006B06BC.sh (in target 'expobug' from project 'expobug')
[RUN_FASTLANE] (1 failure)
[RUN_FASTLANE] Exit status: 65
[RUN_FASTLANE] 
[RUN_FASTLANE] +---------------------------------------+
[RUN_FASTLANE] |           Build environment           |
[RUN_FASTLANE] +-------------+-------------------------+
[RUN_FASTLANE] | xcode_path  | /Applications/Xcode.app |
[RUN_FASTLANE] | gym_version | 2.221.1                 |
[RUN_FASTLANE] | sdk         | iPhoneOS17.5.sdk        |
[RUN_FASTLANE] +-------------+-------------------------+
[RUN_FASTLANE] Looks like fastlane ran into a build/archive error with your project
[RUN_FASTLANE] It's hard to tell what's causing the error, so we wrote some guides on how
[RUN_FASTLANE] to troubleshoot build and signing issues: https://docs.fastlane.tools/codesigning/getting-started/
[RUN_FASTLANE] Before submitting an issue on GitHub, please follow the guide above and make
[RUN_FASTLANE] sure your project is set up correctly.
[RUN_FASTLANE] fastlane uses `xcodebuild` commands to generate your binary, you can see the
[RUN_FASTLANE] the full commands printed out in yellow in the above log.
[RUN_FASTLANE] Make sure to inspect the output above, as usually you'll find more error information there
[RUN_FASTLANE] [!] Error building the application - see the log above
[RUN_FASTLANE] Error: The "Run fastlane" step failed with an unknown error. Refer to "Xcode Logs" below for additional, more detailed logs.

Build failed
The "Run fastlane" step failed with an unknown error. Refer to "Xcode Logs" below for additional, more detailed logs.
npx exited with non-zero code: 1
    Error: build command failed.

```

I managed to track down the issue and narrow it to Metro Bundler:

```
Starting Metro Bundler
warning: Bundler cache is empty, rebuilding (this may take a minute)
/private/var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/0a6b6099-5d23-4241-b7f1-6ba9a0f0d638/build/.:0: error: Unable to resolve module /var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/0a6b6099-5d23-4241-b7f1-6ba9a0f0d638/build/index.js from /private/var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/0a6b6099-5d23-4241-b7f1-6ba9a0f0d638/build/.: 
Error: Unable to resolve module /var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/0a6b6099-5d23-4241-b7f1-6ba9a0f0d638/build/index.js from /private/var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/0a6b6099-5d23-4241-b7f1-6ba9a0f0d638/build/.: 

None of these files exist:
  * ../../../../../../../../../var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/0a6b6099-5d23-4241-b7f1-6ba9a0f0d638/build/index.js(.ios.ts|.native.ts|.ts|.ios.tsx|.native.tsx|.tsx|.ios.mjs|.native.mjs|.mjs|.ios.js|.native.js|.js|.ios.jsx|.native.jsx|.jsx|.ios.json|.native.json|.json|.ios.cjs|.native.cjs|.cjs|.ios.scss|.native.scss|.scss|.ios.sass|.native.sass|.sass|.ios.css|.native.css|.css)
  * ../../../../../../../../../var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/0a6b6099-5d23-4241-b7f1-6ba9a0f0d638/build/index.js/index(.ios.ts|.native.ts|.ts|.ios.tsx|.native.tsx|.tsx|.ios.mjs|.native.mjs|.mjs|.ios.js|.native.js|.js|.ios.jsx|.native.jsx|.jsx|.ios.json|.native.json|.json|.ios.cjs|.native.cjs|.cjs|.ios.scss|.native.scss|.scss|.ios.sass|.native.sass|.sass|.ios.css|.native.css|.css)
Error: Unable to resolve module /var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/0a6b6099-5d23-4241-b7f1-6ba9a0f0d638/build/index.js from /private/var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/0a6b6099-5d23-4241-b7f1-6ba9a0f0d638/build/.: 

None of these files exist:
  * ../../../../../../../../../var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/0a6b6099-5d23-4241-b7f1-6ba9a0f0d638/build/index.js(.ios.ts|.native.ts|.ts|.ios.tsx|.native.tsx|.tsx|.ios.mjs|.native.mjs|.mjs|.ios.js|.native.js|.js|.ios.jsx|.native.jsx|.jsx|.ios.json|.native.json|.json|.ios.cjs|.native.cjs|.cjs|.ios.scss|.native.scss|.scss|.ios.sass|.native.sass|.sass|.ios.css|.native.css|.css)
  * ../../../../../../../../../var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/0a6b6099-5d23-4241-b7f1-6ba9a0f0d638/build/index.js/index(.ios.ts|.native.ts|.ts|.ios.tsx|.native.tsx|.tsx|.ios.mjs|.native.mjs|.mjs|.ios.js|.native.js|.js|.ios.jsx|.native.jsx|.jsx|.ios.json|.native.json|.json|.ios.cjs|.native.cjs|.cjs|.ios.scss|.native.scss|.scss|.ios.sass|.native.sass|.sass|.ios.css|.native.css|.css)
    at ModuleResolver.resolveDependency (/private/var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/0a6b6099-5d23-4241-b7f1-6ba9a0f0d638/build/node_modules/metro/src/node-haste/DependencyGraph/ModuleResolution.js:112:15)
    at DependencyGraph.resolveDependency (/private/var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/0a6b6099-5d23-4241-b7f1-6ba9a0f0d638/build/node_modules/metro/src/node-haste/DependencyGraph.js:231:43)
    at /private/var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/0a6b6099-5d23-4241-b7f1-6ba9a0f0d638/build/node_modules/metro/src/lib/transformHelpers.js:156:21
    at Server._resolveRelativePath (/private/var/folders/38/b0bm_bgj6_l5kt18mth2_8n00000gn/T/eas-build-local-nodejs/0a6b6099-5d23-4241-b7f1-6ba9a0f0d638/build/node_modules/metro/src/Server.js:1023:12)
``` 

### Reproducible demo or steps to reproduce from a blank project

1. git clone https://github.com/krystianhub/ExpoEASBug
2. cd ExpoEASBug
3. npm install
4. **eas build --profile production --platform ios --local**
5. Local build fails
---
1. **eas build --profile production --platform ios**
2. EAS Cloud build is successful
---
1. **eas build --profile development --platform ios --local**
3. Local development build is successful

**Comment by esm-kawamura** [comment]
Same thing happened to me after upgrading to SDK51

**Comment by TakaoNarikawa** [comment]
Same issue occurs when building for both iOS and Android. However, I found that there is no issue when using `fastlane gym` or `./gradlew bundleRelease`.

**Comment by zollipaul** [comment]
Same thing is happening to me with eas local build. The eas cloud build is working though.

**Comment by FlandiaYingman** [comment]
Downgrading from 51 to Expo SDK 50 solves this problem.

**Comment by esm-kawamura** [comment]
I changed  main prop in package.json to "../index.js", eas local build was successful.

**Comment by jvgeee** [comment]
Seeing this too with local builds on EAS.

- `eas build --profile preview --platform ios --local` === FAIL (Local build)
- `eas build --profile preview --platform ios` === PASS (cloud build)
- Xcode archive build === PASS

Building directly in Xcode as an archive build and then submitting to Testflight works just fine, so my guess is that this is a direct issue with EAS CLI local builds

**Comment by jvgeee** [comment]
UPDATE: 

@krystianhub @zollipaul  / anyone else who has this issue, this isn't a direct fix, but downgrading Expo to `51.0.14` solved this issue for me.

Obviously something broke on Expo's end between `51.0.14` and `51.0.17`, so this issue is still relevant. Also ensure that in your `package.json` expo is set to `51.0.14` and not `~51.0.14` as the latter will resolve to a newer (broken) version.

**Comment by szdziedzic** [comment]
Hi, thanks for the report! 

We could reproduce the issue locally on one Expo team member's machine. Interestingly, the repro app was built successfully on another dev's machine, so it seems to be some strange setup-related issue. 

We are trying to find the root cause of it.

**Comment by szdziedzic** [comment]
The issue you are seeing might be related to https://github.com/expo/expo/issues/30295

**Comment by jvgeee** [comment]
@szdziedzic I tried a few different builds with different `main` options, including:

- `"main": "node_modules/expo/AppEntry.js"` => This is what I normally use
- `"main": "App.tsx"` => my app entry file
- `"main": "index.js"` => swapped my app entry file over to `index.js` just in case typescript was getting in the way

None of these worked, builds all failed. 

With `51.0.14` using `"main": "node_modules/expo/AppEntry.js"` works fine so I've kept it at that.

**Comment by janwiebe-jump** [comment]
I've had such issue last week.
After changing `"main": "node_modules/expo/AppEntry.js"` to `"main": "expo/AppEntry.js"` it suddenly worked.

**Comment by keepersmith** [comment]
In case this helps someone else...
I was having this exact same problem (command: eas build --profile test --platform ios --local --clear-cache), and after trying a million different things, it suddenly started working after upgrading expo to 51.0.20.
So at least for me, this issue went away with 51.0.20.

**Comment by byCedric** [comment]
That's because the fix expo/expo#30340 was actually released in `expo@51.0.20`. If you are running into this, upgrade to the latest SDK 51 version :)

**Comment by aravind-enadhu** [comment]
Same issue happens in Expo 55 latest and downgrading to Expo 54 resolves the issue.
